cmd/snap-confine/snap-device-helper-test.c

```diff
@@ -0,0 +1,198 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "../libsnap-confine-private/test-utils.h"
+
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <fcntl.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <string.h>
+
+// TODO: build at runtime
+static const char *sdh_path_default = "snap-confine/snap-device-helper";
+
+// A variant of unsetenv that is compatible with GDestroyNotify
+static void my_unsetenv(const char *k)
+{
+	g_unsetenv(k);
+}
+
+// A variant of rm_rf_tmp that calls g_free() on its parameter
+static void rm_rf_tmp_free(gchar * dirpath)
+{
+	rm_rf_tmp(dirpath);
+	g_free(dirpath);
+}
+
+static gchar *find_sdh_path(void)
+{
+	const char *sdh_from_env = g_getenv("SNAP_DEVICE_HELPER");
+	if (sdh_from_env != NULL) {
+		return g_strdup(sdh_from_env);
+	}
+	return g_strdup(sdh_path_default);
+}
+
+static int run_sdh(gchar * action,
+		   gchar * appname, gchar * devpath, gchar * majmin)
+{
+	gchar *mod_appname = g_strdup(appname);
+	gchar *sdh_path = find_sdh_path();
+
+	// snap-device-helper expects appname to ba passed as <snap>_<app>
+	for (size_t i = 0; i < strlen(mod_appname); i++) {
+		if (mod_appname[i] == '.') {
+			mod_appname[i] = '_';
+			break;
+		}
+	}
+	g_debug("appname modified from %s to %s", appname, mod_appname);
+
+	GError *err = NULL;
+
+	GPtrArray *argv = g_ptr_array_new();
+	g_ptr_array_add(argv, sdh_path);
+	g_ptr_array_add(argv, action);
+	g_ptr_array_add(argv, mod_appname);
+	g_ptr_array_add(argv, devpath);
+	g_ptr_array_add(argv, majmin);
+	g_ptr_array_add(argv, NULL);
+
+	int status = 0;
+
+	gboolean ret = g_spawn_sync(NULL, (gchar **) argv->pdata, NULL, 0,
+				    NULL, NULL, NULL, NULL, &status, &err);
+	g_free(mod_appname);
+	g_free(sdh_path);
+	g_ptr_array_unref(argv);
+
+	if (!ret) {
+		g_debug("failed with: %s", err->message);
+		g_error_free(err);
+		return -2;
+	}
+
+	return WEXITSTATUS(status);
+}
+
+struct sdh_test_data {
+	char *action;
+	char *app;
+	char *file_with_data;
+	char *file_with_no_data;
+};
+
+static void test_sdh_action(gconstpointer test_data)
+{
+	struct sdh_test_data *td = (struct sdh_test_data *)test_data;
+
+	gchar *mock_dir = g_dir_make_tmp(NULL, NULL);
+	g_assert_nonnull(mock_dir);
+	gchar *app_dir = g_build_filename(mock_dir, td->app, NULL);
+	gchar *with_data = g_build_filename(mock_dir,
+					    td->app,
+					    td->file_with_data,
+					    NULL);
+	gchar *without_data = g_build_filename(mock_dir,
+					       td->app,
+					       td->file_with_no_data,
+					       NULL);
+	int ret = 0;
```

Comment:
Why is "ret" declared here instead of below in run_sdh? The other vars (like gchar *whiteout_data) get declared where they are used. Similar question for data.

Reply:
Indeed, I'll fix `ret`. As for `data`, it's later used in `g_file_get_contents(with_data, &data, NULL, NULL)`